bug: The mysqld server crash when a left join and group by query is executed at Large amount of data and disk In abundance

### Have you read the Contributing Guidelines on issues?

- [X] I have read the [Contributing Guidelines on issues](https://github.com/stoneatom/stonedb/blob/stonedb-5.7-dev/CONTRIBUTING.md).

### Please confirm if bug report does NOT exists already ?

- [X] I confirm there is no existing issue for this

### Describe the problem

```sql
create table A(id int,UnitSN varchar(10),CTIME datetime);
insert into A values(1,'A1','2022-12-09 01:00:00');
insert into A values(2,'A2','2022-12-09 02:00:00');
insert into A values(3,'A3','2022-12-09 03:00:00');
insert into A values(4,'A4','2022-12-09 04:00:00');
insert into A values(5,'A5','2022-12-09 05:00:00');
insert into A values(6,'A6','2022-12-09 06:00:00');
insert into A values(7,'A7','2022-12-09 07:00:00');
insert into A values(8,'A8','2022-12-09 08:00:00');
insert into A values(9,'A9','2022-12-09 09:00:00');
insert into A values(10,'A10','2022-12-09 10:00:00');
insert into A values(11,'A11','2022-12-09 11:00:00');
insert into A values(12,'A12','2022-12-09 12:00:00');
insert into A values(13,'A13','2022-12-09 13:00:00');
insert into A values(14,'A14','2022-12-09 14:00:00');
insert into A values(15,'A15','2022-12-09 15:00:00');
insert into A values(16,'A16','2022-12-09 16:00:00');
insert into A values(17,'A17','2022-12-09 17:00:00');
insert into A values(18,'A18','2022-12-09 18:00:00');
insert into A values(19,'A19','2022-12-09 19:00:00');
insert into A values(20,'A20','2022-12-09 20:00:00');
insert into A select * from A;
......
mysql> select count(*) from A;
+----------+
| count(*) |
+----------+
|    20480 |
+----------+
1 row in set (0.00 sec)

create table B(id int,AID int,Name varchar(10),value varchar(10));
insert into B values(1,1,'V1','XXXXXXX1');
insert into B values(2,1,'V2','XXXXXXX2');
insert into B values(3,1,'V3','XXXXXXX3');
insert into B values(4,1,'V4','XXXXXXX4');
insert into B values(5,1,'V5','XXXXXXX5');
insert into B values(6,2,'V6','XXXXXXX6');
insert into B values(7,2,'V7','XXXXXXX7');
insert into B values(8,2,'V8','XXXXXXX8');
insert into B values(9,2,'V9','XXXXXXX9');
insert into B values(10,2,'V10','XXXXXXX10');
insert into B values(11,3,'V11','XXXXXXX11');
insert into B values(12,3,'V12','XXXXXXX12');
insert into B values(13,3,'V13','XXXXXXX13');
insert into B values(14,3,'V14','XXXXXXX14');
insert into B values(15,3,'V15','XXXXXXX15');
insert into B values(16,4,'V16','XXXXXXX16');
insert into B values(17,4,'V17','XXXXXXX17');
insert into B values(18,4,'V18','XXXXXXX18');
insert into B values(19,4,'V19','XXXXXXX19');
insert into B values(20,4,'V20','XXXXXXX20');
insert into B values(21,5,'V21','XXXXXXX21');
insert into B values(22,5,'V22','XXXXXXX22');
insert into B values(23,5,'V23','XXXXXXX23');
insert into B values(24,5,'V24','XXXXXXX24');
insert into B values(25,5,'V25','XXXXXXX25');
insert into B values(26,6,'V26','XXXXXXX26');
insert into B values(27,6,'V27','XXXXXXX27');
insert into B values(28,6,'V28','XXXXXXX28');
insert into B values(29,6,'V29','XXXXXXX29');
insert into B values(30,6,'V30','XXXXXXX30');
insert into B select * from B;
......
mysql> select count(*) from B;
+----------+
| count(*) |
+----------+
| 10149888 |
+----------+
1 row in set (0.01 sec)

###crash SQL
select a.ID,
       a.UNITSN,
       case upper(b.NAME) when 'V1' then b.VALUE else null end as V1,
       case upper(b.NAME) when 'V2' then b.VALUE else null end as V2,
       case upper(b.NAME) when 'V3' then b.VALUE else null end as V3,
       case upper(b.NAME) when 'V4' then b.VALUE else null end as V4,
       case upper(b.NAME) when 'V5' then b.VALUE else null end as V5
  from A a
  left join B b
    on a.ID = b.AID
 group by a.ID, a.UNITSN;
```

```sql
(gdb) bt
#0  __pthread_kill_implementation (no_tid=0, signo=6, threadid=139737402230336) at ./nptl/pthread_kill.c:44
#1  __pthread_kill_internal (signo=6, threadid=139737402230336) at ./nptl/pthread_kill.c:78
#2  __GI___pthread_kill (threadid=139737402230336, signo=signo@entry=6) at ./nptl/pthread_kill.c:89
#3  0x00007f172cb44476 in __GI_raise (sig=sig@entry=6) at ../sysdeps/posix/raise.c:26
#4  0x00007f172cb2a7f3 in __GI_abort () at ./stdlib/abort.c:79
#5  0x00007f172cb2a71b in __assert_fail_base (fmt=0x7f172ccdf150 "%s%s%s:%u: %s%sAssertion `%s' failed.\n%n",
    assertion=0x55d6b81e6c26 "n < size", file=0x55d6b81e6bf0 "/stonedb/storage/tianmu/core/index_table.h", line=41,
    function=<optimized out>) at ./assert/assert.c:92
#6  0x00007f172cb3be96 in __GI___assert_fail (assertion=0x55d6b81e6c26 "n < size",
    file=0x55d6b81e6bf0 "/stonedb/storage/tianmu/core/index_table.h", line=41,
    function=0x55d6b81e6bb0 "void Tianmu::core::IndexTable::Set64(uint64_t, uint64_t)") at ./assert/assert.c:101
#7  0x000055d6b7b65789 in Tianmu::core::IndexTable::Set64 (this=0x7f1258914150, n=649592832, val=1281)
    at /stonedb/storage/tianmu/core/index_table.h:41
#8  0x000055d6b7b8ac9c in Tianmu::core::MultiIndexBuilder::BuildItem::CommitTableValues (this=0x7f1258937960)
    at /stonedb/storage/tianmu/core/multi_index_builder.cpp:88
#9  0x000055d6b7b94423 in Tianmu::core::ParallelHashJoiner::SubmitJoinedTuple (this=0x7f1258934730, build_item=0x7f1258937960,
    traversed_hash_table=0x7f12580273c0, hash_row=2528968, mit=...) at /stonedb/storage/tianmu/core/parallel_hash_join.cpp:1049
#10 0x000055d6b7b93d06 in Tianmu::core::ParallelHashJoiner::AsyncMatchDim (this=0x7f1258934730, params=0x7f1258935680)
    at /stonedb/storage/tianmu/core/parallel_hash_join.cpp:968
#11 0x000055d6b7b92b4a in Tianmu::core::ParallelHashJoiner::MatchDim (this=0x7f1258934730, mit=...)
    at /stonedb/storage/tianmu/core/parallel_hash_join.cpp:825
#12 0x000055d6b7b8ffae in Tianmu::core::ParallelHashJoiner::ExecuteJoin (this=0x7f1258934730)
    at /stonedb/storage/tianmu/core/parallel_hash_join.cpp:439
#13 0x000055d6b7b8eb51 in Tianmu::core::ParallelHashJoiner::ExecuteJoinConditions (this=0x7f1258934730, cond=...)
    at /stonedb/storage/tianmu/core/parallel_hash_join.cpp:230
#14 0x000055d6b7af4b10 in Tianmu::core::ParameterizedFilter::UpdateJoinCondition (this=0x7f12589270b0, cond=..., tips=...)
    at /stonedb/storage/tianmu/core/parameterized_filter.cpp:651
#15 0x000055d6b7af8e76 in Tianmu::core::ParameterizedFilter::UpdateMultiIndex (this=0x7f12589270b0, count_only=false, limit=-1)
    at /stonedb/storage/tianmu/core/parameterized_filter.cpp:1304
#16 0x000055d6b7719059 in Tianmu::core::Query::Preexecute (this=0x7f1726387f50, qu=..., sender=0x7f125891b730, display_now=true)
    at /stonedb/storage/tianmu/core/query.cpp:795
#17 0x000055d6b76e78de in Tianmu::core::Engine::Execute (this=0x55d6b9bb3370, thd=0x7f1258004d60, lex=0x7f1258007088,
    result_output=0x7f1258026ed0, unit_for_union=0x0) at /stonedb/storage/tianmu/core/engine_execute.cpp:477
#18 0x000055d6b76e6647 in Tianmu::core::Engine::HandleSelect (this=0x55d6b9bb3370, thd=0x7f1258004d60, lex=0x7f1258007088,
    result=@0x7f17263883e8: 0x7f1258026ed0, setup_tables_done_option=0, res=@0x7f17263883dc: 32530,
    optimize_after_tianmu=@0x7f17263883e4: 1, tianmu_free_join=@0x7f17263883e0: 1, with_insert=0)
    at /stonedb/storage/tianmu/core/engine_execute.cpp:238
--Type <RET> for more, q to quit, c to continue without paging--
#19 0x000055d6b78001de in Tianmu::handler::ha_my_tianmu_query (thd=0x7f1258004d60, lex=0x7f1258007088,
    result_output=@0x7f17263883e8: 0x7f1258026ed0, setup_tables_done_option=0, res=@0x7f17263883dc: 32530,
    optimize_after_tianmu=@0x7f17263883e4: 1, tianmu_free_join=@0x7f17263883e0: 1, with_insert=0)
    at /stonedb/storage/tianmu/handler/ha_my_tianmu.cpp:88
#20 0x000055d6b6d838b2 in execute_sqlcom_select (thd=0x7f1258004d60, all_tables=0x7f1258026888) at /stonedb/sql/sql_parse.cc:5217
#21 0x000055d6b6d7c8dd in mysql_execute_command (thd=0x7f1258004d60, first_level=true) at /stonedb/sql/sql_parse.cc:2856
#22 0x000055d6b6d84940 in mysql_parse (thd=0x7f1258004d60, parser_state=0x7f1726389590) at /stonedb/sql/sql_parse.cc:5655
#23 0x000055d6b6d7947f in dispatch_command (thd=0x7f1258004d60, com_data=0x7f1726389e40, command=COM_QUERY)
    at /stonedb/sql/sql_parse.cc:1495
#24 0x000055d6b6d782f6 in do_command (thd=0x7f1258004d60) at /stonedb/sql/sql_parse.cc:1034
#25 0x000055d6b6ec0c28 in handle_connection (arg=0x55d6bf0154f0) at /stonedb/sql/conn_handler/connection_handler_per_thread.cc:313
#26 0x000055d6b760ce89 in pfs_spawn_thread (arg=0x55d6bee76730) at /stonedb/storage/perfschema/pfs.cc:2197
#27 0x00007f172cb96b43 in start_thread (arg=<optimized out>) at ./nptl/pthread_create.c:442
#28 0x00007f172cc28a00 in clone3 () at ../sysdeps/unix/sysv/linux/x86_64/clone3.S:81

```

### Expected behavior

_No response_

### How To Reproduce

_No response_

### Environment

./mysqld  Ver 5.7.36-StoneDB-debug for Linux on x86_64 (build-)
build information as follow:
        Repository address: https://github.com/stoneatom/stonedb.git:stonedb-5.7-dev
        Branch name: stonedb-5.7-dev
        Last commit ID: 2a03448e7
        Last commit time: Date:   Tue Dec 6 10:35:03 2022 +0800
        Build time: Date: Thu 08 Dec 2022 12:45:27 PM UTC

### Are you interested in submitting a PR to solve the problem?

- [X] Yes, I will!

### When viewing logs, a disk insufficiency error message is displayed

```bash

[2022-12-14 22:22:12.107476] [2617] [ERROR] [parallel_hash_join.cpp:812] MSG: An exception is caught: I/O operation on cache/JW_INT00000108167E340x7fd4c4.tianmu_tmp failed with error - No space left on device
[2022-12-14 22:22:12.107499] [2617] [ERROR] [parallel_hash_join.cpp:812] MSG: An exception is caught: I/O operation on cache/JW_INT00000131EBD6450x7fd4c4.tianmu_tmp failed with error - No space left on device
[2022-12-14 22:22:12.107512] [2617] [ERROR] [parallel_hash_join.cpp:812] MSG: An exception is caught: I/O operation on cache/JW_INT0000011090E6D60x7fd4c4.tianmu_tmp failed with error - No space left on device


```

--- a/storage/tianmu/core/index_table.cpp
+++ b/storage/tianmu/core/index_table.cpp
@@ -121,15 +121,15 @@ void IndexTable::LoadBlock(int b) {
   block_changed = false;
 }
 
-void IndexTable::ExpandTo(int64_t new_size) {
+void IndexTable::ExpandTo(uint64_t new_size) {
   DEBUG_ASSERT(IsLocked());
-  if (new_size <= (int64_t)size)
+  if (new_size <= (uint64_t)size)
     return;
   if (size * bytes_per_value == buffer_size_in_bytes) {  // the whole table was in one buffer
     if (buffer_size_in_bytes < 32_MB && size_t(new_size * bytes_per_value) > 32_MB) {
       max_buffer_size_in_bytes =
-          int(mm::TraceableObject::MaxBufferSize(-1));   // recalculate, as it might not be done earlier
-      CI_SetDefaultSize((int)max_buffer_size_in_bytes);  // redefine disk block sized
+          uint64_t(mm::TraceableObject::MaxBufferSize(-1));   // recalculate, as it might not be done earlier
+      CI_SetDefaultSize((uint64_t)max_buffer_size_in_bytes);  // redefine disk block sized
       uint values_per_block = uint(max_buffer_size_in_bytes / bytes_per_value);
       block_shift =
           CalculateBinSize(values_per_block) - 1;  // e.g. BinSize(16)==5, but shift by 4. WARNING: should it be
@@ -137,9 +137,9 @@ void IndexTable::ExpandTo(int64_t new_size) {
       block_mask = (uint64_t(1) << block_shift) - 1;
     }
 
-    int new_buffer_size_in_bytes;
-    if (new_size * bytes_per_value < (int64_t)max_buffer_size_in_bytes)
-      new_buffer_size_in_bytes = int(new_size * bytes_per_value);
+    uint64_t new_buffer_size_in_bytes;
+    if (new_size * bytes_per_value < (uint64_t)max_buffer_size_in_bytes)
+      new_buffer_size_in_bytes = uint64_t(new_size * bytes_per_value);
     else
       new_buffer_size_in_bytes = max_buffer_size_in_bytes;
     // TODO: check the rc_alloc status

--- a/storage/tianmu/core/index_table.h
+++ b/storage/tianmu/core/index_table.h
@@ -108,7 +108,7 @@ class IndexTable : private system::CacheableItem, public mm::TraceableObject {
                                             // which may occur in table
   uint64_t N() { return size; }             // note: this is the upper size, the table can be used partially!
   int BlockShift() { return block_shift; }  // block = int( tuple >> block_shift )
-  void ExpandTo(int64_t new_size);
+  void ExpandTo(uint64_t new_size);
 
   // mm::TraceableObject functionality
   mm::TO_TYPE TraceableType() const override { return mm::TO_TYPE::TO_INDEXTABLE; }
@@ -118,10 +118,10 @@ class IndexTable : private system::CacheableItem, public mm::TraceableObject {
 
   unsigned char *buf = nullptr;  // polymorphic: unsigned short, unsigned int or int64_t
 
-  int max_buffer_size_in_bytes;
+  uint64_t max_buffer_size_in_bytes;
   int bytes_per_value;
   int max_block_used;
-  size_t buffer_size_in_bytes;
+  uint64_t buffer_size_in_bytes;
   int block_shift;
   uint64_t block_mask;
   uint64_t size;

--- a/storage/tianmu/core/multi_index_builder.h
+++ b/storage/tianmu/core/multi_index_builder.h
@@ -47,7 +47,7 @@ class MultiIndexBuilder {
 
     IndexTable *ReleaseIndexTable(int dim);
     bool NullExisted(int dim) const { return nulls_possible_[dim]; }
-    int GetCount() const { return added_count_; }
+    uint64_t GetCount() const { return added_count_; }
     bool IsEmpty() const { return (added_count_ == 0) ? true : false; }
 
    private:
@@ -59,7 +59,7 @@ class MultiIndexBuilder {
     std::vector<int64_t> cached_values_;
     std::vector<bool> nulls_possible_;
 
-    int added_count_ = 0;
+    uint64_t added_count_ = 0;
     IndexTable **index_table_;
     std::unique_ptr<MINewContentsRSorter> rough_sorter_;
   };

--- a/storage/tianmu/system/cacheable_item.h
+++ b/storage/tianmu/system/cacheable_item.h
@@ -51,7 +51,7 @@ class CacheableItem {
   // or its subrange from position 'off', of length 'size' bytes.
   // Returns 0 if successfully got the (sub)block, -1 otherwise
 
-  void CI_SetDefaultSize(int size)  // If the value is set (here, or in constructor),
+  void CI_SetDefaultSize(uint64_t size)  // If the value is set (here, or in constructor),
   {
     default_block_size_ = size;
   }  // then the size parameter in CI_Put may be omitted.
@@ -73,7 +73,7 @@ class CacheableItem {
   std::vector<int> file_start_;     // an address in the file where the i-th data block starts
   std::vector<int> file_size_;      // a size of the i-th data block
 
-  int default_block_size_ = 0;
+  uint64_t default_block_size_ = 0;
   TianmuFile cur_file_handle_;
   int cur_file_number_ = -1;  // the number of currently opened file
   void *cur_map_addr_ = nullptr;

